datalad/distribution/update.py

```diff
@@ -355,73 +443,127 @@ def _choose_merge_target(repo, branch, remote, cfg_remote):
         # branch.*.merge value, but that assumes a value for remote.*.fetch.
-        merge_target = repo.call_git_oneline(
+        target = repo.call_git_oneline(
             ["rev-parse", "--symbolic-full-name", "--abbrev-ref=strict",
              "@{upstream}"],
             read_only=True)
     elif branch:
         remote_branch = "{}/{}".format(remote, branch)
         if repo.commit_exists(remote_branch):
-            merge_target = remote_branch
-    return merge_target
+            target = remote_branch
+    return target
 
 
-#  Merge functions
+#  Update functions
 
 
-def _choose_merge_fn(repo, is_annex=False, adjusted=False):
-    if adjusted and is_annex:
-        # For adjusted repos, blindly sync.
-        merge_fn = _annex_sync
-    elif is_annex:
-        merge_fn = _annex_plain_merge
-    elif adjusted:
+def _choose_update_fn(repo, how, is_annex=False, adjusted=False):
+    if adjusted and how != "merge":
         raise RuntimeError(
-            "bug: Upstream checks should make it impossible for "
-            "adjusted=True, is_annex=False")
+            "bug: Upstream checks should abort if adjusted is used "
+            "with action other than 'merge'")
+    elif how in ["merge", "ff-only"]:
+        if adjusted and is_annex:
+            # For adjusted repos, blindly sync.
+            fn = _annex_sync
+        elif is_annex:
+            fn = _annex_plain_merge
+        elif adjusted:
+            raise RuntimeError(
+                "bug: Upstream checks should make it impossible for "
+                "adjusted=True, is_annex=False")
+        else:
+            fn = _plain_merge
+    elif how == "reset":
+        fn = _reset_hard
+    elif how == "checkout":
+        fn = _checkout
     else:
-        merge_fn = _plain_merge
-    return merge_fn
+        raise ValueError(f"Unrecognized value for `how`: {how}")
+    return fn
+
+
+def _try_command(record, fn, *args, **kwargs):
+    """Call `fn`, catching a `CommandError`.
 
+    Parameters
+    ----------
+    record : dict
+        A partial result record. It should at least have 'action' and 'message'
+        fields. A 'status' value of 'ok' or 'error' will be added based on
+        whether calling `fn` raises a `CommandError`.
 
-def _plain_merge(repo, _, target, merge_opts=None):
+    Returns
+    -------
+    A new record with a 'status' field.
+    """
     try:
-        repo.merge(name=target, options=merge_opts,
-                   expect_fail=True, expect_stderr=True)
+        fn(*args, **kwargs)
     except CommandError as exc:
-        yield {"action": "merge", "status": "error",
-               "message": exc_str(exc)}
+        # TODO: "message" is dropped from the json{_pp} representation. Either
+        # that should change or the exc_str result should be put somewhere
+        # else. Could be "exception_traceback" (see get_status_dict), but then
+        # that's not visible with default renderer.
```

Comment:
With the merge of gh-5536, the message would be retained in the `json{,_pp}` results.

Reply:
Merged now.